src/irc/mod.rs

```diff
@@ -47,48 +47,61 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
     server_model: models::ServerModel,
 }
 
-#[derive(Debug, Clone, Default)]
+#[derive(Clone)]
 struct UserNickBuilder {
+    ctx: ConnectionContext,
     nick: Option<String>,
     user: Option<String>,
     real_name: Option<String>,
     password: Option<String>,
 }
+impl UserNickBuilder {
+    fn with_context(ctx: ConnectionContext) -> Self {
+        UserNickBuilder {
+            ctx,
+            nick: None,
+            user: None,
+            real_name: None,
+            password: None,
+        }
+    }
+}
 
 impl UserNickBuilder {
     fn to_user_nick(self) -> UserNick {
         UserNick {
             nick: self.nick.unwrap(),
             user: self.user.unwrap(),
             real_name: self.real_name.unwrap(),
-            password: self.password.unwrap(),
+            password: self.password,
         }
     }
+    /// Check that the user has input all fields except password.
+    ///
+    /// A missing password will be caught in `await_login`.
```

Comment:
Is this change in behavior somehow linked to porting to async/await? If not, please do this in a separate PR, this one is already very large.

Reply:
Im not sure what you mean. The changes regarding getting IRC user information from the irc connection stream required a large rewrite of `stream_fold.rs` + the creation of `stream_fold::StateUpdate` trait to be compatible with the `Pin<&mut Self>` requirements of futures 0.3.

As far as checking the password field in `await_login`, this is an addition that I made recently to mirror the way futures 0.1 does it after spotting the discrepancy in the diffs:

https://github.com/matrix-org/matrix-ircd/blob/7d5f660e11037ee380a08ed74cb6ea95dba3bb9d/src/irc/mod.rs#L96-L104

Reply:
How would you like to proceed?

Reply:
Sounds like it is indeed related, so consider this resolved.